### PR TITLE
Cypress Includes for External Validation

### PR DESCRIPTION
**Pull request by conroywhitney**
In order to allow projects to validate their cat3 files against a cypress ProductTest's expected results, we need to make those libraries available outside of the cypress web app. One approach could be creating a gem which encapsulates this, which could be an approach down the line. Another could be setting up a specific web endpoint which would do the validation on an external platform; however, that could make automated testing brittle should the host be unresponsive.

This approach simply allows the validating project to include the `lib/includes_for_external_validation.rb` file and have access to all of the classes and libraries which are used for validation.

### CODE REVIEW

File: lib/includes_for_external_validation.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  include Cypress::Validators
+end
+
+# load HQMF from health-data-standards' lib folder
+Gem.find_files('hqmf-parser.rb').each { |f| require f }
+
+APP_CONFIG = {
+  'file_upload_root' => File.absolute_path('./tmp'),
+  'effective_date'   => {
+    'year' => '2013'
```
Comment (dillonwelch) [inline lib/includes_for_external_validation.rb:84]: Seems like something we'd want to allow the caller to configure if possible.

File: lib/validators/qrda_cat3_validator.rb [hunk 2]
```diff
@@ -1,4 +1,3 @@
-require 'validators/qrda_cat1_validator'
```
Comment (dillonwelch) [inline lib/validators/qrda_cat3_validator.rb:1]: Was this require removed because it causes errors, or just because it's unnecessary?

File: app/models/calculated_product_test.rb [hunk 3]
```diff
@@ -54,14 +54,15 @@ def generate_records
   end
 
   def validators(doc)
-    @validators ||= [::Validators::QrdaCat3Validator.new(expected_results),
-      ::Validators::MeasurePeriodValidator.new(),
-      ::Validators::ExpectedResultsValidator.new(expected_results)]
+    @validators ||= [
+      ::Validators::QrdaCat3Validator.new(expected_results),
+      ::Validators::MeasurePeriodValidator.new,
+      ::Validators::ExpectedResultsValidator.new(expected_results)
```
Comment (dillonwelch) [inline app/models/calculated_product_test.rb]: Doesn't look like there's any changes here outside of Rubocop issues? If so, can you revert to keep the branching from upstream small?

File: app/models/calculated_product_test.rb [hunk 4]
```diff
@@ -129,4 +130,15 @@ def update_new_result_ids(res, qrda_id, new_rec_id)
     res_clone.save
   end
 
+  def read_qrda_file_contents(qrda_file)
+    qrda_file.respond_to?(:open) ? qrda_file.open.read : qrda_file.read
+  end
+
+  def should_generate_population?
+    !population_already_generated?
```
Comment (dillonwelch) [inline app/models/calculated_product_test.rb:136]: Fold `population_already_generated?` into here?

File: lib/includes_for_external_validation.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
+# gems
+require 'aasm'
+require 'devise'
+require 'devise/orm/mongoid'
+require 'carrierwave'
+require 'carrierwave/mongoid'
+require 'health-data-standards'
+require 'health-data-standards/hqmf-parser'
```
Comment (dillonwelch) [inline lib/includes_for_external_validation.rb]: I thought this didn't do anything?

File: lib/includes_for_external_validation.rb [hunk 6]
```diff
@@ -0,0 +1,76 @@
+# gems
+require 'aasm'
+require 'devise'
+require 'devise/orm/mongoid'
+require 'carrierwave'
+require 'carrierwave/mongoid'
+require 'health-data-standards'
```
Comment (dillonwelch) [inline lib/includes_for_external_validation.rb]: Since HDS + QME are in the Gemfile here + full dependencies of our app, can we remove these requires?

File: lib/includes_for_external_validation.rb [hunk 7]
```diff
@@ -0,0 +1,76 @@
+# gems
+require 'aasm'
+require 'devise'
```
Comment (dillonwelch) [inline lib/includes_for_external_validation.rb:3]: Odd that we'd need devise if we're not doing any controllers.

**Comment by dillonwelch** [comment]
Overall, this looks nice and the approach I was thinking of. 

I didn't see the WIP again this time :+1:

File: lib/validators/qrda_cat3_validator.rb (hunk 2 above)
Comment (conroywhitney) [inline lib/validators/qrda_cat3_validator.rb:1]: Both.  :smile: 

It caused errors because it was not a require_relative and therefore was being required from the project using Cypress for external validation.

We also already require that file in `includes_for_external_validation.rb` and Cypress was still green after removing it.

**tl;dr** Not sure why it was in there to begin with. Doesn't seem to be necessary.

File: app/models/calculated_product_test.rb (hunk 4 above)
Comment (conroywhitney) [inline app/models/calculated_product_test.rb:136]: I considered that approach (it actually used to all be inline!), but in lieu of comments explaining what we were doing and why, I went with the more verbose methods to explicitly explain to future me why and how we were bypassing automatic population generation.

If it's OK by you, I'd like to leave this.

File: lib/includes_for_external_validation.rb (hunk 7 above)
Comment (conroywhitney) [inline lib/includes_for_external_validation.rb:3]: I know, it's less than ideal  :confused: 

This is actually necessary w.r.t. loading `Cypress::User` fixtures in our 3rd party project which is using Cypress for external validation. For the same reason, we `add_development_dependency 'devise'` in our other project's `Gemfile` and need to set all sorts of `devise`-related fields on our `Cypress::User` fixtures.

This was just a first-pass "can this even work?" approach. Definitely open to suggestions. I'd love to use a different approach if you can think of one.

File: lib/includes_for_external_validation.rb (hunk 1 above)
Comment (conroywhitney) [inline lib/includes_for_external_validation.rb:84]: The 3rd party project using Cypress for external validation should be able to override the `APP_CONFIG` key to update their year in their `spec/support/cypress.rb` file.

Just as a side note, though, this date is what is hard-coded in Cypress' `config/cypress.yml`

**Comment by conroywhitney** [comment]
:+1: I should have removed the WIP after our out-of-band conversation. Thanks for all your feedback!

What's next for this PR? Do some of the above comments still need to be addressed? Are we going to merge this into master, or keep it on its own separate branch for a while until we're 110% sure of this direction? Questions? Concerns? Suggestions?

**Comment by dillonwelch** [comment]
Keeping as a separate branch sounds like a good idea to keep updates easier. Also, it's not as important to keep updated since it's not a runtime dependency for our API.

I will respond to above comments today.

File: app/models/calculated_product_test.rb (hunk 4 above)
Comment (dillonwelch) [inline app/models/calculated_product_test.rb:136]: :+1:

File: lib/validators/qrda_cat3_validator.rb (hunk 2 above)
Comment (dillonwelch) [inline lib/validators/qrda_cat3_validator.rb:1]: :+1:

File: lib/includes_for_external_validation.rb (hunk 1 above)
Comment (dillonwelch) [inline lib/includes_for_external_validation.rb:84]: :+1:
